### DISCUSSION TITLE
Using optional filters together with compiled query

### DISCUSSION THREAD

**Discussion by sebarys** [slick/slick#2314] [Questions] [answered]
Hi,

Is it possible to use optional filter values in `Compiled` query? Kind of `filterOpt` inside compiled query.

Context: I have query with ~30 filters that can be optionally applied as `filterOpt` is not supported in Compiled queries I'm not able to compile it. Part of query:
```
      Tables.MyTable
        .filter(_.name === nameValue)
        .filterOpt(queryParams.active)(_.active === _)
        .filterOpt(queryParams.paused)(_.paused === _)
        .filterOpt(queryParams.status)(_.status inSetBind _)
        .filterOpt(queryParams.createdAt)(_.createdAt <= _)
        .filterOpt(queryParams.modifiedAt)(_.modifiedAt >= _)
        ...
```

I guess it would need create sth like:
```
(OptionalParam IS NULL OR filtered_column = OptionalParam)
```

Should I use different function inside Compiled query to achieve such behaviour?

Thanks for any suggestions :)

**Comment by nafg** [comment]
Can you show the code with Compiled and the error?

Otherwise you could try `table.filter(t => value.map(t.col === _).getOrElse(LiteralColumn(true)))`

**Comment by sebarys** [comment]
Thanks for info

Code with filterOpt:
```
  private lazy val queryCompiled = Compiled { (id: Rep[String], active: Rep[Option[Boolean]]) =>
    Tables.Users
      .filter(_.id === id)
      .filterOpt(active)(_.active === _)
  }
```

Compilation error:
```
type mismatch;
 found   : com.my.project.DatabaseProfile.api.Rep[Option[Boolean]]
    (which expands to)  slick.lifted.Rep[Option[Boolean]]
 required: Option[?]
      .filterOpt(active)(_.active === _)
```

I've tried alternative but Rep[Option] is not supporting `.getOrElse(LiteralColumn(true)))` only `get` is supported and throw exception if value is `None`.

I've also tried 
```
  private lazy val queryCompiled = Compiled { (id: Rep[String], active: Rep[Option[Boolean]]) =>
    Tables.Users
      .filter(_.relyingParty === id)
      .filter(users => if (active.isDefined) active.map(users.active === _) else true)
  }
```

but again issue with `Rep` 
```
type mismatch;
 found   : slick.lifted.Rep[Boolean]
 required: Boolean
      .filter(users => if (active.isDefined) active.map(users.active === _) else true)
```

**Comment by nafg** [comment, marked as answer]
> Thanks for info
> 
> Code with filterOpt:
> 
> ```
>   private lazy val queryCompiled = Compiled { (id: Rep[String], active: Rep[Option[Boolean]]) =>
>     Tables.Users
>       .filter(_.id === id)
>       .filterOpt(active)(_.active === _)
>   }
> ```
> 
> Compilation error:
> 
> ```
> type mismatch;
>  found   : com.my.project.DatabaseProfile.api.Rep[Option[Boolean]]
>     (which expands to)  slick.lifted.Rep[Option[Boolean]]
>  required: Option[?]
>       .filterOpt(active)(_.active === _)
> ```

What if you define active as `Option[Boolean]` without being a Rep?


> 
> I've tried alternative but Rep[Option] is not supporting `.getOrElse(LiteralColumn(true)))` only `get` is supported and throw exception if value is `None`.

It should if you wildcard import from api


> 
> I've also tried
> 
> ```
>   private lazy val queryCompiled = Compiled { (id: Rep[String], active: Rep[Option[Boolean]]) =>
>     Tables.Users
>       .filter(_.relyingParty === id)
>       .filter(users => if (active.isDefined) active.map(users.active === _) else true)
>   }
> ```
> 
> but again issue with `Rep`
> 
> ```
> type mismatch;
>  found   : slick.lifted.Rep[Boolean]
>  required: Boolean
>       .filter(users => if (active.isDefined) active.map(users.active === _) else true)
> ```

Because scala's `if` keyword needs an actual Boolean, of course.

Whenever you see Rep, that means it's a value that does not exist in the JVM, instead it's a typed fragment out of which to build SQL, so it's only going to be evaluated inside the database. So Scala code cannot inspect a Rep or do different behavior depending on its value. The only thing you can do with a Rep is build more SQL out of it. Slick provides extension methods on different types of Reps that return more Reps. So for example, unlike getOrElse on Option which actually decides what to do based on the actual Option value it's called on, Slick provides getOrElse (IIRC) etc. on `Rep[Option[...]]` that generates SQL to do a similar thing to regular getOrElse but that will of course be evaluated within the database.

(Technically, a Rep does not contain generated SQL, instead Slick generates SQL by inspecting Reps)

**Reply by sebarys** [reply]
Tanks for clarification around `Ref` - it was helpful.

> What if you define active as Option[Boolean] without being a Rep?

I think it is not possible for the same reasons that you've described about using Ref inside `if` condition etc.

Anyway `getOrElse` is working, thanks for pointing out this, I'm not able to find what import exactly I'm missing but `import slick.lifted._` solve issue :) 


Working query:
```
  lazy val queryCompiled = Compiled { (id: Rep[String], active: Rep[Option[Boolean]]) =>
    Tables.Users
      .filter(_.id === id)
      .filter(user => active.map(user.active === _).getOrElse(LiteralColumn(true)))
  }
```
produce SQL:
```
select "id", "some_column", "another_column", "active"
from "users"
where ("id" = ?)
    and coalesce(
        (case when (? is null) then null else ("active" = ?) end),
        true
    )
```

**Comment by wuservices** [comment]
I was following this because I ran into a similar issue.

However I originally was using `filter` with `Compiled`, which generated a logically equivalent query in the database. In my case this was MySQL, so I switched to using `filterOpt` without compilation to simplify the generated query because MySQL wasn't using an index for the optional field when a `CASE` was added to the generated SQL.

Thanks for the `Rep` explanation. That makes sense. It sounds like once you have a `Rep` for `Compiled`, it's not possible to make a decision outside the database? Would one solution be to just have 2 compiled queries and then you choose which one to run?

I guess another option would be to use plain SQL.

**Reply by nafg** [reply]
Sure, you could have 2 compiled queries.

How necessary for you is it to use compiled queries altogether, though?

**Reply by nafg** [reply]
Anyway, `getOrElse` is definitely defined for `Option[Rep[...]]` so there should not be an issue using it if you import `api._`

**Reply by sebarys** [reply]
> How necessary for you is it to use compiled queries altogether, though?

In production grade systems, shouldn't all executed queries be compiled?
Do you know any way to discover overhead connected with generating final query? Ideally I would like to know how long it take to generate SQL for given query, so I can decide whether I want to spend time to make query compiled or I can accept runtime overhead on each execution.
